scripts/build.py

```diff
@@ -142,50 +239,43 @@ def build_files(source, target):
-    BUILD_DIR = os.path.join('build', CACHE_SLUG)
-
-    # os.path.dirname(path)(in ensure_directory_exists()) returns parent
-    # directory of a path passed as an argument to it. This is as intended
-    # for file paths, but for directory paths we do not want this to happen,
-    # hence we append a trailing slash to it.
-    # Process assets, copy it to build/[cache_slug]/assets
-    ASSETS_SRC_DIR = os.path.join('assets', '')
-    ASSETS_OUT_DIR = os.path.join(BUILD_DIR, 'assets', '')
-    copy_files_source_to_target(ASSETS_SRC_DIR, ASSETS_OUT_DIR)
+    hashes = dict() # pylint: disable=C0103
 
-    # Process third_party resources, copy it to
-    # build/[cache_slug]/third_party/generated
+    # Create hashes for assets, copy folders and files to build/assets
+    ASSETS_SRC_DIR = os.path.join('assets', '')
+    ASSETS_OUT_DIR = os.path.join('build', 'assets', '')
+    hashes.update(get_file_hashes(ASSETS_SRC_DIR))
+    copy_files_source_to_target(ASSETS_SRC_DIR, ASSETS_OUT_DIR, hashes)
+
+    # Process third_party resources, create hashes for them and copy them into
+    # build/third_party/generated
+    THIRD_PARTY_GENERATED_PROD_DIR = os.path.join(
+        'third_party', 'generated', '')
     THIRD_PARTY_GENERATED_OUT_DIR = os.path.join(
-        BUILD_DIR, 'third_party', 'generated')
-    build_minified_third_party_libs(THIRD_PARTY_GENERATED_OUT_DIR)
-
-    # Minify extension static resources, copy it to
-    # build/[cache_slug]/extensions
-    EXTENSIONS_SRC_DIR = os.path.join('extensions', '')
-    EXTENSIONS_OUT_DIR = os.path.join(BUILD_DIR, 'extensions', '')
-    build_files(EXTENSIONS_SRC_DIR, EXTENSIONS_OUT_DIR)
-
-    TEMPLATES_HEAD_DIR = os.path.join('core', 'templates', 'dev', 'head', '')
-    TEMPLATES_OUT_DIR = os.path.join('core', 'templates', 'prod', 'head', '')
-    build_files(TEMPLATES_HEAD_DIR, TEMPLATES_OUT_DIR)
-
-    # Process core/templates/prod/head/css, copy it to build/[cache_slug]/css
-    CSS_SRC_DIR = os.path.join('core', 'templates', 'prod', 'head', 'css', '')
-    CSS_OUT_DIR = os.path.join(BUILD_DIR, 'css', '')
-    copy_files_source_to_target(CSS_SRC_DIR, CSS_OUT_DIR)
-
-    # Copy core/templates/prod/head/ to build/[cache_slug]/templates/head/
-    TEMPLATES_SRC_DIR = os.path.join('core', 'templates', 'prod', 'head', '')
-    TEMPLATES_OUT_DIR = os.path.join(BUILD_DIR, 'templates', 'head', '')
-    copy_files_source_to_target(TEMPLATES_SRC_DIR, TEMPLATES_OUT_DIR)
+        'build', 'third_party', 'generated', '')
+    build_minified_third_party_libs(THIRD_PARTY_GENERATED_PROD_DIR)
+    hashes.update(get_file_hashes(THIRD_PARTY_GENERATED_PROD_DIR))
+    copy_files_source_to_target(
+        THIRD_PARTY_GENERATED_PROD_DIR, THIRD_PARTY_GENERATED_OUT_DIR, hashes)
+
+    # Minify extension static resources, create hashes for them and copy them
+    # into build/extensions
+    EXTENSIONS_DEV_DIR = os.path.join('extensions', '')
+    EXTENSIONS_PROD_DIR = os.path.join('extensions', 'prod', '')
```

Comment:
I don't understand, why do we have an EXTENSIONS_PROD_DIR? Is this a temporary dir for staging?

If so, can we have a /build_staging dir where we do all this stuff so that we don't have random prod dirs all over the codebase?